src/libcore/iter/mod.rs

```diff
@@ -1954,6 +2293,40 @@ impl<I> Iterator for Take<I> where I: Iterator{
 
         (lower, upper)
     }
+
+    #[inline]
+    fn try_fold<Acc, Fold, R>(&mut self, init: Acc, mut fold: Fold) -> R where
+        Self: Sized, Fold: FnMut(Acc, Self::Item) -> R, R: Try<Ok=Acc>
+    {
+        if self.n == 0 {
+            Try::from_ok(init)
+        } else {
+            let n = &mut self.n;
+            self.iter.try_fold(init, move |acc, x| {
+                *n -= 1;
+                let r = fold(acc, x);
+                if *n == 0 { SearchResult::Found(r) }
+                else { SearchResult::from_try(r) }
+            }).into_try()
+        }
+    }
+
+    #[inline]
+    fn fold<Acc, Fold>(mut self, init: Acc, mut fold: Fold) -> Acc
+        where Fold: FnMut(Acc, Self::Item) -> Acc,
+    {
+        let mut n = self.n;
+        if n == 0 {
+            init
+        } else {
+            self.iter.try_fold(init, move |acc, x| {
+                n -= 1;
+                let acc = fold(acc, x);
+                if n == 0 { SearchResult::Found(acc) }
+                else { SearchResult::NotFound(acc) }
+            }).into_inner()
+        }
```

Comment:
can't this `fold` just use `.try_fold()`?

Reply:
I was worried about `SearchResult::from_try(r)` potentially pessimizing things, since it does a bunch of destruct-rebuild. 
 There may be no basis behind that, though, since it's not the more important case of "fold should use fold to pick up any existing overrides".  I'll just delete it, since less code is better :)